### DISCUSSION TITLE
Local installation of visualizer possible?

### DISCUSSION THREAD

**Discussion by raydeleu** [sgenoud/replicad#7] [Q&A] [answered]
Is it possible to install the visualizer for Replicad locally, either using the live webserver in VSCode or as a progressive web application (wpa)?

**Comment by sgenoud** [comment, marked as answer]
Good point, I have quickly added the functionality! Don't hesitate to report if you see any issue with it!

**Reply by raydeleu** [reply]
This works perfectly! Thank you for the quick solution. 

Some remarks that might be helpful: 

- It is now very easy to install the Visualizer, create some code in a local text editor and feed this to the Visualizer. The behaviour is then very similar to using Cadquery or CascadeStudio. 
- A big difference to Cadquery and CascadeStudio is that the error messages in the Visualizer are difficult to read due to the small window for these messages. 
- Furthermore it is difficult to understand the error that was made. Perhaps this would be solved when installing a local version of Replicad in a webserver, but this is not a simple task.  
- The Visualizer has a very neat inertia effect when changing the viewing position, but for modeling purposes this is distracting. As a user I would prefer to exactly determine the viewpoint. 
- The methods offered by Replicad to identify edges and faces is complex. It is similar to Cadquery but an approach to select the extremities of the shape (without determining the exact distance) is missing. CascadeStudio has an approach where the user is supported in the viewer to identify the index of the edges and faces, which helps in complex cases. The downside of using edge and face indices instead of the methods currently offered is that this is less robust in case of design changes. Offering both would offer a beginning user the advantage of being able to work out if the error in the code is due to the edge selection or due to kernel constraints. 
- I did not find a way to visualize two shapes. From the examples I gather that it is possible, but I was not able to work out how. 
- It it is possible to display several shapes, consider the option to colour them differently
- Would it be possible to influence the resolution of an STL export? 
- For a beginner the documentation is difficult to understand. I would expect a structure similar to that of the tutorial (which is very nice!) 

All in all I get the impression that Replicad is aimed mostly at a use case where a model is prepared by an expert and then  transferred to a website. The website is then only used to tweak design parameters to adjust the offered product. As such it is possibly not comparable to OpenSCAD, Cadquery or CascadeStudio that function as a CAD program with a coding interface.

**Reply by sgenoud** [reply]
thanks for the feedback!

> All in all I get the impression that Replicad is aimed mostly at a use case where a model is prepared by an expert and then transferred to a website.

 I developed replicad to allow me to build website to generate 3D models - and it feels like it shows on its design. The visualizer has been so far mostly an afterthought to showcase what it possible.

But I am interested to make it better. Would you be willing to stick to it a bit more while I improve things around?

> Furthermore it is difficult to understand the error that was made. Perhaps this would be solved when installing a local version of Replicad in a webserver, but this is not a simple task.

Definitely an area that can be improved. Most of the problem comes from a performance-nice error message tradeoff. I am thinking of some ways I could make it better - don't hesitate to create issues when you get into a "Kernel Error" and you don't know what the problem is.

> Offering both would offer a beginning user the advantage of being able to work out if the error in the code is due to the edge selection or due to kernel constraints.

I have created issues to add these kinds of selections

> The Visualizer has a very neat inertia effect when changing the viewing position, but for modeling purposes this is distracting. As a user I would prefer to exactly determine the viewpoint.

I have already received a similar feedback - I guess my taste is in the minority. It is now changed

> I did not find a way to visualize two shapes. From the examples I gather that it is possible, but I was not able to work out how.

if you return an array of shapes, there is a dropdown on the left to select which one you want. I will add another options to show more than one in the same screen (and have them with different colors).

> Would it be possible to influence the resolution of an STL export?

Sure!

> For a beginner the documentation is difficult to understand. I would expect a structure similar to that of the tutorial (which is very nice!)

I have seen what you have written for CascadeStudio. When I improve the visualiser to something better, would you be willing to adapt it to create a "guide" in the official documentation?

Thanks again for the feedback!

**Reply by raydeleu** [reply]
I noticed the improvement in the Visualizer and already checked the display of multiple objects. This works as you describe, but I was more looking for the option to show them together, like the example of the "baby trainer" that has two objects in the scene. Cadquery offers something like an assembly. In OpenCascade this would probably be achieved in a "Compound" of solids. 

If I end up regularly using a tool I am certainly willing to put some time in. I think that I am the biggest user of my own manual :-). Having some interaction and getting help is certainly a big motivator for improving things.

**Reply by sgenoud** [reply]
Some more improvements:

<img width="391" alt="Capture d’écran 2022-06-17 à 16 07 46" src="https://user-images.githubusercontent.com/263325/174314734-65dda1fa-40f2-4de9-8a36-8e596a5a9852.png">

You now have a "All shapes" option to see...all shapes at once. And you can set the base color of your shapes:

```js
const main = ({ draw, drawRoundedRectangle }) => {
  const shape = drawRoundedRectangle(20, 30, 5).sketchOnPlane().extrude(10);

  return [
    { shape: shape, name: "purple shape", color: "purple" },
    {
      shape: shape.clone().translateZ(15),
      color: "steelblue",
      name: "blue shape",
    },
    {
      shape: shape.clone().translateZ(30),
      color: "orange",
      name: "orange shape",
    },
  ];
};

```
